pychebfun/chebfun.py

```diff
@@ -156,7 +156,10 @@ def sample_function(self, f, N):
         Sample a function on N+1 Chebyshev points.
         """
         x = self.interpolation_points(N+1)
-        return f(x)
+        try:
+            return f(x)
+        except:  # needed when trying to sample wrapped C functions
```

Comment:
Could you add the proper exception? Is it `ArgumentError`?

Reply:
Hmm I understand your point but the problem is that `ArgumentError` is defined within the **ctypes** module. I can't tag it by name without importing it from that module else I get a `NameError` saying "name 'ArgumentError' is not defined". Do you want me to import this and re-implement?

Reply:
Yes, please try that. It would be ideal if you could also write a test that makes sure that the error is fixed (based on the original text of the PR, for instance).

Reply:
The problem with doing that is illustrated by the following attempt at trying to approximate the C function using CFFI instead:

```
In [1]: from cffi import FFI

In [2]: ffi = FFI()

In [3]: ffi.cdef("double myfunc(double t);")

In [4]: C = ffi.dlopen("libtest.so")

In [5]: C.myfunc(0.1)
Out[5]: 1.3968022466674206

In [6]: from pychebfun import *
   ...: cheb = chebfun(C.myfunc)
   ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-7807e3879a0e> in <module>
      1 from pychebfun import *
----> 2 cheb = chebfun(C.myfunc)

/usr/local/lib/python3.6/dist-packages/pychebfun/chebfun.py in chebfun(f, domain, N, chebcoeff)
    334     # callable
    335     if hasattr(f, '__call__'):
--> 336         return Chebfun.from_function(f, domain, N)
    337 
    338     # from here on, assume that f is None, or iterable

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in from_function(self, f, domain, N)
    129 
    130         # Find out the right number of coefficients to keep
--> 131         coeffs = self.dichotomy(**args)
    132 
    133         return self.from_coeff(coeffs, domain)

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in dichotomy(self, f, kmin, kmax, raise_no_convergence)
     94             N = pow(2, k)
     95 
---> 96             sampled = self.sample_function(f, N)
     97             coeffs = self.polyfit(sampled)
     98 

/usr/local/lib/python3.6/dist-packages/pychebfun/chebfun.py in sample_function(self, f, N)
    157         """
    158         x = self.interpolation_points(N+1)
--> 159         return f(x)
    160 
    161     @classmethod

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in <lambda>(t)
    119         a,b = domain[0], domain[-1]
    120         map_ui_ab = lambda t: 0.5*(b-a)*t + 0.5*(a+b)
--> 121         args = {'f': lambda t: f(map_ui_ab(t))}
    122         if N is not None: # N is provided
    123             nextpow2 = int(np.log2(N))+1

TypeError: only size-1 arrays can be converted to Python scalars
```

Then you have Boost::Python, swig, Cython… I didn't even try to wrap those as I don't have (enough) experience with them.

I understand that your concern is that errors to the wrapping which may occur within the `f(x)` call shouldn't be caught by the `except` clause, but consider this: the except clause still calls `f(x)`, except it just does that via the `vectorize` wrapper. Any errors that would have occured at the first call from `try` will occur again from the `except` clause, and this time there is no safety net so the errors will propagate.

So I think the PR as it stands is OK and doesn't constitute a problem. Maybe I can add a comment for future readers of the code saying that this `try`-`except` is only for supporting functions (mostly C ones) which can't take an `ndarray` and any other errors within the evaluation of `f(x)` will be caught the second time around. But that should be enough.

Reply:
To illustrate:

```
In [1]: from math import sin

In [2]: def mysine(i):
   ...:     return sin(i) + "hello"
   ...: 

In [3]: from pychebfun import chebfun

In [4]: cheb = chebfun(mysine)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.6/dist-packages/pychebfun/chebfun.py in sample_function(self, f, N)
    159         try:
--> 160             return f(x)
    161         except:  # needed when trying to sample wrapped C functions

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in <lambda>(t)
    120         map_ui_ab = lambda t: 0.5*(b-a)*t + 0.5*(a+b)
--> 121         args = {'f': lambda t: f(map_ui_ab(t))}
    122         if N is not None: # N is provided

<ipython-input-2-42df05c39787> in mysine(i)
      1 def mysine(i):
----> 2     return sin(i) + "hello"

TypeError: only size-1 arrays can be converted to Python scalars

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-4-4a99b379e21c> in <module>
----> 1 cheb = chebfun(mysine)

/usr/local/lib/python3.6/dist-packages/pychebfun/chebfun.py in chebfun(f, domain, N, chebcoeff)
    337     # callable
    338     if hasattr(f, '__call__'):
--> 339         return Chebfun.from_function(f, domain, N)
    340 
    341     # from here on, assume that f is None, or iterable

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in from_function(self, f, domain, N)
    129 
    130         # Find out the right number of coefficients to keep
--> 131         coeffs = self.dichotomy(**args)
    132 
    133         return self.from_coeff(coeffs, domain)

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in dichotomy(self, f, kmin, kmax, raise_no_convergence)
     94             N = pow(2, k)
     95 
---> 96             sampled = self.sample_function(f, N)
     97             coeffs = self.polyfit(sampled)
     98 

/usr/local/lib/python3.6/dist-packages/pychebfun/chebfun.py in sample_function(self, f, N)
    160             return f(x)
    161         except:  # needed when trying to sample wrapped C functions
--> 162             return np.vectorize(f)(x)
    163 
    164     @classmethod

/usr/local/lib/python3.6/dist-packages/numpy/lib/function_base.py in __call__(self, *args, **kwargs)
   2089             vargs.extend([kwargs[_n] for _n in names])
   2090 
-> 2091         return self._vectorize_call(func=func, args=vargs)
   2092 
   2093     def _get_ufunc_and_otypes(self, func, args):

/usr/local/lib/python3.6/dist-packages/numpy/lib/function_base.py in _vectorize_call(self, func, args)
   2159             res = func()
   2160         else:
-> 2161             ufunc, otypes = self._get_ufunc_and_otypes(func=func, args=args)
   2162 
   2163             # Convert args to object arrays first

/usr/local/lib/python3.6/dist-packages/numpy/lib/function_base.py in _get_ufunc_and_otypes(self, func, args)
   2119 
   2120             inputs = [arg.flat[0] for arg in args]
-> 2121             outputs = func(*inputs)
   2122 
   2123             # Performance note: profiling indicates that -- for simple

/usr/local/lib/python3.6/dist-packages/pychebfun/polyfun.py in <lambda>(t)
    119         a,b = domain[0], domain[-1]
    120         map_ui_ab = lambda t: 0.5*(b-a)*t + 0.5*(a+b)
--> 121         args = {'f': lambda t: f(map_ui_ab(t))}
    122         if N is not None: # N is provided
    123             nextpow2 = int(np.log2(N))+1

<ipython-input-2-42df05c39787> in mysine(i)
      1 def mysine(i):
----> 2     return sin(i) + "hello"

TypeError: unsupported operand type(s) for +: 'float' and 'str'
```

Reply:
I agree, in fact, the error will come from any function which is not vectorizable, whatever the reason.

Reply:
Please add the following test in `test_chebfun.py`:
```python
def test_vectorize():
    def not_vectorized(x):
        if x > 0:
            return 1
        else:
            return 0
    Chebfun.sample_function(not_vectorized, 10)
```
and change the docstring/comment, not mentioning sage or C functions. Any functions are now acceptable with your change.

Reply:
To make sure that exception raised by the function are not caught (as you showed above), you can add the following test as well to `test_chebfun.py`:
```python
def test_sample_leaves_errors():
    """
    Chebfun.sample does not catch errors from the sampling function
    """
    def silly(x):
        return 1/0
    with pytest.raises(ZeroDivisionError):
        Chebfun.sample_function(silly, 10)
```

Reply:
Um not to be naïve but are these to be added at the top level of the module or as member functions of the unittest classes therein?

Reply:
As separate functions (top level).